src/libraries/System.Net.WebSockets/tests/WebSocketTestStream.cs

```diff
@@ -0,0 +1,240 @@
+using System.Threading.Tasks;
+
+namespace System.Net.WebSockets.Tests
+{
+    /// <summary>
+    /// A helper stream class that can be used simulate sending / receiving (duplex) data in a websocket.
+    /// </summary>
+    public class WebSocketTestStream : Stream
+    {
+        private readonly SemaphoreSlim _inputLock = new(initialCount: 0);
+        private readonly Queue<Block> _inputQueue = new();
+        private readonly CancellationTokenSource _disposed = new();
+
+        public WebSocketTestStream()
+        {
+            GC.SuppressFinalize(this);
+            Remote = new WebSocketTestStream(this);
+        }
+
+        private WebSocketTestStream(WebSocketTestStream remote)
+        {
+            GC.SuppressFinalize(this);
+            Remote = remote;
+        }
+
+        public WebSocketTestStream Remote { get; }
+
+        /// <summary>
+        /// Returns the number of unread bytes.
+        /// </summary>
+        public int Available
+        {
+            get
+            {
+                var available = 0;
+
+                lock (_inputQueue)
+                {
+                    foreach (var x in _inputQueue)
+                    {
+                        available += x.AvailableLength;
+                    }
+                }
+
+                return available;
+            }
+        }
+
+        public Span<byte> NextAvailableBytes
+        {
+            get
+            {
+                lock (_inputQueue)
+                {
+                    if (_inputQueue.TryPeek(out Block block))
+                    {
+                        return block.Available;
+                    }
+                    return default;
+                }
+            }
+        }
+
+        /// <summary>
+        /// If set, would cause the next send operation to be delayed
+        /// and complete asynchronously. Can be used to test cancellation tokens
+        /// and async code branches.
+        /// </summary>
+        public TimeSpan DelayForNextSend { get; set; }
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => true;
+
+        public override long Length => -1;
+
+        public override long Position { get => -1; set => throw new NotSupportedException(); }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!_disposed.IsCancellationRequested)
+            {
+                _disposed.Cancel();
+
+                lock (Remote._inputQueue)
+                {
+                    Remote._inputLock.Release();
+                    Remote._inputQueue.Enqueue(Block.ConnectionClosed);
+                }
+            }
+        }
+
+        public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            using (var cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _disposed.Token))
+            {
```

Comment:
A using declaration can reduce the indent hill, and `var` shouldn't be used unless the RHS has a `new`.

```suggestion
            using CancellationTokenSource cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _disposed.Token));
```